File: frontend/src/services/api/purchaseOrder.ts
import api from '../axios';

export interface PurchaseOrderItem {
  id?: string;
  product_id: string;
  product_code?: string;
  product_name?: string;
  ordered_quantity: number;
  received_quantity?: number;
  unit_price: number;
  subtotal?: number;
  status?: string;
}

export interface PurchaseOrder {
  id: string;
  po_number: string;
  supplier: string;
  status: string;
  total_amount: number;
  expected_date?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
  items: PurchaseOrderItem[];
}

export interface CreatePurchaseOrderRequest {
  supplier: string;
  expected_date?: string;
  notes?: string;
  items: Array<{
    product_id: string;
    ordered_quantity: number;
    unit_price: number;
  }>;
}

export const purchaseOrderAPI = {
  // 발주서 목록 조회
  getAll: (params?: { status?: string; supplier?: string }) => {
    return api.get<PurchaseOrder[]>('/purchase-orders', { params });
  },

  // 발주서 상세 조회
  getById: (id: string) => {
    return api.get<PurchaseOrder>(`/purchase-orders/${id}`);
  },

  // 발주서 생성
  create: (data: CreatePurchaseOrderRequest) => {
    return api.post<PurchaseOrder>('/purchase-orders', data);
  },

  // 발주서 상태 업데이트
  updateStatus: (id: string, status: string) => {
    return api.put(`/purchase-orders/${id}/status`, null, { params: { status } });
  },

  // 입고 처리
  receiveItems: (id: string, items: Array<{ item_id: string; quantity: number }>) => {
    return api.post(`/purchase-orders/${id}/receive`, items);
  }
};

File: backend/app/models/purchase_order.py
"""
Purchase Order Models
"""
from sqlalchemy import Column, String, Integer, ForeignKey, Text, Date, Numeric
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.models.base import BaseModel


class PurchaseOrder(BaseModel):
    """발주서 테이블"""
    __tablename__ = "purchase_orders"
    
    # 발주서 정보
    po_number = Column(String(50), unique=True, nullable=False, index=True)
    supplier = Column(String(200))
    
    # 상태: draft(임시저장), waiting(입고대기), partial(부분입고), completed(완료)
    status = Column(String(20), default='draft', index=True)
    
    # 금액 및 날짜
    total_amount = Column(Numeric(12, 2))
    expected_date = Column(Date)
    
    # 기타
    notes = Column(Text)
    created_by = Column(String(100))
    
    # 관계
    items = relationship("PurchaseOrderItem", back_populates="purchase_order", cascade="all, delete-orphan")
    
    def __repr__(self):
        return f"<PurchaseOrder {self.po_number}: {self.status}>"


class PurchaseOrderItem(BaseModel):
    """발주 상세 테이블"""
    __tablename__ = "purchase_order_items"
    
    # 발주서 참조
    po_id = Column(UUID(as_uuid=True), ForeignKey('purchase_orders.id'), nullable=False, index=True)
    
    # 제품 정보
    product_code = Column(String(50), ForeignKey('products.product_code'), nullable=False, index=True)
    
    # 수량 및 가격
    ordered_quantity = Column(Integer, nullable=False)
    received_quantity = Column(Integer, default=0)
    unit_price = Column(Numeric(12, 2))
    subtotal = Column(Numeric(12, 2))
    
    # 상태: pending(대기), partial(부분입고), completed(완료)
    status = Column(String(20), default='pending')
    
    # 관계
    purchase_order = relationship("PurchaseOrder", back_populates="items")
    product = relationship("Product", foreign_keys=[product_code], back_populates="purchase_order_items")
    
    def __repr__(self):
        return f"<POItem PO:{self.po_id} Product:{self.product_code} Qty:{self.ordered_quantity}>"